In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm 
from sklearn.preprocessing import PolynomialFeatures

In [3]:
data = {
    'Size': [1500, 1600, 1700, 1800, 1900, 2000, 2100, 2200],
    'Bedrooms': [3, 3, 3, 4, 4, 4, 5, 5],
    'Age': [10, 15, 10, 20, 30, 15, 20, 25],
    'Price': [300000, 320000, 340000, 360000, 400000, 370000, 420000, 430000]
}

In [9]:
df=pd.DataFrame(data)

In [21]:
X = df[['Size']]
y=df['Price']

In [23]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [25]:
model = LinearRegression()
# Train the model
model.fit(X_train, y_train)
# Make predictions on the test set
y_pred = model.predict(X_test)

In [27]:
print("Intercept:", model.intercept_)
print("Coefficients:", model.coef_)

Intercept: 16600.00000000006
Coefficients: [192.]


In [29]:
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [39]:
print("Mean Squared Error:", mse)
print("R Squared:", r2)

Mean Squared Error: 475400000.0
R Squared: 0.23936000000000002


In [45]:
new_house = np.array([[1800]])
predicted_price = model.predict(new_house)
print(f"Predicted Price for new house: {predicted_price[0]}")

Predicted Price for new house: 362200.0


C:\Users\vani1\anaconda3\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [47]:
df=pd.read_csv("C:\\Users\\vani1\\Downloads\\wine.csv")

In [49]:
df.head()

Wine  Alcohol  Malic.acid   Ash   Acl   Mg  Phenols  Flavanoids  \
0     1    14.23        1.71  2.43  15.6  127     2.80        3.06   
1     1    13.20        1.78  2.14  11.2  100     2.65        2.76   
2     1    13.16        2.36  2.67  18.6  101     2.80        3.24   
3     1    14.37        1.95  2.50  16.8  113     3.85        3.49   
4     1    13.24        2.59  2.87  21.0  118     2.80        2.69   

   Nonflavanoid.phenols  Proanth  Color.int   Hue    OD  Proline  
0                  0.28     2.29       5.64  1.04  3.92     1065  
1                  0.26     1.28       4.38  1.05  3.40     1050  
2                  0.30     2.81       5.68  1.03  3.17     1185  
3                  0.24     2.18       7.80  0.86  3.45     1480  
4                  0.39     1.82       4.32  1.04  2.93      735

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Wine                  178 non-null    int64  
 1   Alcohol               178 non-null    float64
 2   Malic.acid            178 non-null    float64
 3   Ash                   178 non-null    float64
 4   Acl                   178 non-null    float64
 5   Mg                    178 non-null    int64  
 6   Phenols               178 non-null    float64
 7   Flavanoids            178 non-null    float64
 8   Nonflavanoid.phenols  178 non-null    float64
 9   Proanth               178 non-null    float64
 10  Color.int             178 non-null    float64
 11  Hue                   178 non-null    float64
 12  OD                    178 non-null    float64
 13  Proline               178 non-null    int64  
dtypes: float64(11), int64(3)
memory usage: 19.6 KB


In [93]:
X = df[['Alcohol']]
y=df['Wine']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model = LinearRegression()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"R Squared: {r2:.4f}")

R Squared: 0.1304


r2 Scores of features:
alcohol: 0.1304 ; Malic acid: 0.1003 ; ash: -0.0238 ; acl: 0.2347 ; mg: 0.0007 ; phenols: 0.6590 ; Flavanoids: 0.7466 ; Nonflavanoid.phenols : 0.2647
Proanth: 0.2515 ; Color.int:-0.0604 ; Hue: 0.3937 ; OD: 0.5501 ; Proline : 0.4395

Columns to Consider (Strong Predictors)
These columns have relatively high R² values, meaning they contribute significantly to predicting Wine:

Flavanoids: 0.7466 → Strongest predictor
Phenols: 0.6590 → Strong predictor
OD (Optical Density): 0.5501 → Moderate predictor
Proline: 0.4395 → Moderate predictor
Hue: 0.3937 → Moderate predictor
Acl (Alcohol Content Level): 0.2347 → Weak to moderate predictor
These variables should be considered in your model.

Columns to Discard (Weak Predictors)
These columns have very low or negative R² values, meaning they have little to no predictive power:

Color.int: -0.0604 (Negative correlation, discard)
Ash: -0.0238 (Negative correlation, discard)
Mg (Magnesium): 0.0007 (Very close to zero, discard)
Malic.acid: 0.1003 (Weak predictor, discard)
Nonflavanoid.phenols: 0.2647 (Weak predictor, consider removing)
Proanth: 0.2515 (Weak predictor, may remove)
Alcohol: 0.1304 (Weak predictor, consider removing)


Final Decision
Keep: Flavanoids, Phenols, OD, Proline, Hue, Acl
Drop: Color.int, Ash, Mg, Malic.acid, Nonflavanoid.phenols, Proanth, Alcohol

In [96]:
X = df[['Flavanoids','Phenols','OD','Proline','Hue','Acl']]
y=df['Wine']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model = LinearRegression()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"R Squared: {r2:.4f}")

R Squared: 0.8678
